plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt

```diff
@@ -0,0 +1,177 @@
+) : PackageManager(name, analysisRoot, analyzerConfig, repoConfig), CommandLineTool {
+    class Factory : AbstractPackageManagerFactory<Pnpm>("PNPM") {
+        override val globsForDefinitionFiles = listOf("package.json", "pnpm-lock.yaml")
+
+        override fun create(
+            analysisRoot: File,
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) = Pnpm(type, analysisRoot, analyzerConfig, repoConfig)
+    }
+
+    private val handler = PnpmDependencyHandler(this)
+    private val graphBuilder by lazy { DependencyGraphBuilder(handler) }
+    private val packageDetailsCache = mutableMapOf<String, PackageJson>()
+
+    override fun resolveDependencies(definitionFile: File, labels: Map<String, String>): List<ProjectAnalyzerResult> =
+        stashDirectories(definitionFile.resolveSibling("node_modules")).use {
+            resolveDependencies(definitionFile)
+        }
+
+    private fun resolveDependencies(definitionFile: File): List<ProjectAnalyzerResult> {
+        val workingDir = definitionFile.parentFile
+        installDependencies(workingDir)
+
+        val workspaceModuleDirs = getWorkspaceModuleDirs(workingDir)
+        handler.setWorkspaceModuleDirs(workspaceModuleDirs)
+
+        val moduleInfosForScope = Scope.entries.associateWith { scope -> listModules(workingDir, scope) }
+
+        return workspaceModuleDirs.map { projectDir ->
+            val project = parseProject(projectDir.resolve("package.json"), analysisRoot, managerName)
+
+            val scopeNames = Scope.entries.mapTo(mutableSetOf()) { scope ->
+                val scopeName = scope.toString()
+                val qualifiedScopeName = DependencyGraph.qualifyScope(project, scope.toString())
+                val moduleInfo = moduleInfosForScope.getValue(scope).single { it.path == projectDir.absolutePath }
+
+                moduleInfo.getScopeDependencies(scope).forEach { dependency ->
+                    graphBuilder.addDependency(qualifiedScopeName, dependency)
+                }
+
+                scopeName
+            }
+
+            ProjectAnalyzerResult(
+                project = project.copy(scopeNames = scopeNames),
+                packages = emptySet(),
+                issues = emptyList()
+            )
+        }
+    }
+
+    override fun command(workingDir: File?) = if (Os.isWindows) "pnpm.cmd" else "pnpm"
+
+    private fun getWorkspaceModuleDirs(workingDir: File): Set<File> {
+        val json = run(workingDir, "list", "--json", "--only-projects", "--recursive").stdout
+
+        return parsePnpmList(json).mapTo(mutableSetOf()) { File(it.path) }
+    }
+
+    private fun listModules(workingDir: File, scope: Scope): List<ModuleInfo> {
+        val scopeOption = when (scope) {
+            Scope.DEPENDENCIES -> "--prod"
+            Scope.DEV_DEPENDENCIES -> "--dev"
+        }
+
+        val json = run(workingDir, "list", "--json", "--recursive", "--depth", "10000", scopeOption).stdout
```

Comment:
How can we confidently limit the depth to 10000 now if previously it was -1?

Reply:
10000 prints out more than -1

Reply:
Really? Isn't -1 an alias for "unlimited"?

Reply:
It seem like.

Reply:
Did you try out -1? (I wonder why you are so sure that this works)

Reply:
I tried, maybe I made a mistake...

Reply:
With "it seem like" i responded to "Really?"

Reply:
```
pnpm list --help | grep depth
      --depth -1           Display only projects. Useful in a monorepo. `pnpm ls
                           -r --depth -1` lists all projects in a monorepo
      --depth <number>     Max display depth of the dependency tree
      --depth 0            Display only direct dependencies
```

Reply:
> I wonder why you are so sure that this works

Because it's being used in the existing code. But now I realize the use is in a different context that does not apply here.

Anyway, the question remains why it is valid to artificially limit the depth of the dependency tree now at an (arbitrary?) number of 10000. I would have expected no limit at all, or if there was a limit, it should be configurable via a package manager specific option to go into the direction of https://github.com/oss-review-toolkit/ort/issues/5626.

Reply:
To my understanding there is no way to make the command work without a limit. 
So, using a large value in my opinion makes sense. I don't think 10k is too small.

If you don't like it still, please make a proposal.

Reply:
> Because it's being used in the existing code. But now I realize the use is in a different context that does not apply here.

I don't know what context you are talking about. But previously existing code didn't call this command at all.

Reply:
> I don't know what context you are talking about. But previously existing code didn't call this command at all.

I was talking about this line:

https://github.com/oss-review-toolkit/ort/blob/507ee30cea999cb6bfd2cdcc2f5bc8e5e4350dd4/plugins/package-managers/node/src/main/kotlin/Pnpm.kt#L59

It *does* use the `--depth` argument for `pnpm list`, but with -1, and in the context of getting workspaces.

Reply:
> If you don't like it still, please make a counter proposal.

I *already* made two counter proposals: 1. just omit `, "--depth", "10000"` completely, or 2. make the depth value configurable.

Reply:
> I already made two counter proposals:

ok, let me try to explain again why I prefer to not go with these:

1. Does not work because `-1` lists only projects, but not dependencies
2. Making this a) extends the scope of this PR and b) I doubt it is needed, 
    due to the rationale I already provided above, e.g:

> I don't think 10k is too small. In fact it is so large that it is only a theroretical problem IMO. ORT's dependency graph breaks the cycles anyway and I doubt the depth limit could be exceeded without printing cycles.

I still believe this to be a theoretical problem here and 10k works in practically any case. 
Do you fear if some dependencies would still be missing in the output? If this is the case, then let's agree on a higher hard-coded value. But please let's not make this configurable when we can easily avoid it / there is no need.

Reply:
> 1. Does not work because `-1` lists only projects, but not dependencies

I don't get why this is relevant. I was not proposing to use `-1`, nor does omitting `--depth 10000` completely imply a value of `-1` AFAIU (because as the help states, `-1` does *not* mean "unlimited" as I thought before, but it's a special value to "Display only projects").

So can you please explain why omitting `--depth 10000` would not work?

Reply:
> So can you please explain why omitting `--depth 10000` would not work?

Because then it uses the default which is `-1`.  (Which is why it is relevant, e.g.)

> I don't get why this is relevant.

Reply:
Very weird that `pnpm list --help` says "All dependencies are printed by default", but the [docs](https://pnpm.io/cli/list#--depth-number) say "`pnpm ls --depth 0` (default) will list direct dependencies only".

However, apparently we can use "Infinity" as a depth value instead, which worked in my tests. Please use that value as it's less arbitrary and documents the intention.

Reply:
`Infinity` is of course much nicer than `10000`. How did you find that one?

Reply:
It's in the linked docs.

Reply:
argh